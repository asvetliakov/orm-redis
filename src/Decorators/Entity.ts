import { REDIS_ENTITY } from "../Metadata/Metadata";

/**
 * Defines redis entity class
 * 
 * @export
 * @param {string} [name] 
 * @returns void
 */
export function Entity(name?: string): ClassDecorator {
    return function (constructor: Function) {
        Reflect.defineMetadata(REDIS_ENTITY, name ? name : constructor.name, constructor);
    };
}