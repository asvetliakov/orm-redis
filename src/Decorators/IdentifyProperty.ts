import { PropertyMetadata, REDIS_PROPERTIES } from "../Metadata/Metadata";

/**
 * Defines identify property for given redis hash.
 * Hash must contain one IdentifyProperty
 * 
 * @export
 * @param type Type
 * @param [name] Redis property name
 * @returns 
 */
export function IdentifyProperty(nameOrType?: string | typeof Number | typeof String, type?: typeof Number | typeof String): PropertyDecorator {
    return function (target: Object, propertyKey: string): void {
        const designType: any = Reflect.getMetadata("design:type", target, propertyKey);
        const name = typeof nameOrType === "string" ? nameOrType : propertyKey;
        const propertyType = typeof nameOrType === "function"
            ? nameOrType
            : type
                ? type
                : designType;        

        if (propertyType !== String && propertyType !== Number) {
            throw new Error("Identify property must be string or number type");
        }
        const properties: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, target.constructor) || [];
        properties.push({
            propertyName: propertyKey,
            propertyRedisName: name,
            isIdentifyColumn: true,
            propertyType: propertyType,
            isRelation: false
        });
        Reflect.defineMetadata(REDIS_PROPERTIES, properties, target.constructor);
    };
}