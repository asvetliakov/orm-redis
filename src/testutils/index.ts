export * from "./redis";
export * from "./RedisTestMonitor";