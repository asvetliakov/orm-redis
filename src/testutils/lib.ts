// See https://github.com/Microsoft/TypeScript/issues/16872
// USE ONLY IN TESTS

const MyMap = Map;
const MySet = Set;
const MyInt8Array = Int8Array;
const MyDate = Date;

export { MyMap as Map };
export { MySet as Set };
export { MyInt8Array as Int8Array };
export { MyDate as Date };