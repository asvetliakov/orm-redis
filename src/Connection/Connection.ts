import { AlreadyConnectedError } from "../Errors/Errors";
import { RedisManager } from "../Persistence/RedisManager";
import { EntitySubscriberInterface } from "../Subscriber/EntitySubscriberInterface";
import { PubSubSubscriberInterface } from "../Subscriber/PubSubSubscriberInterface";
import { getFromContainer } from "../utils/Container";
import * as redis from "../utils/PromisedRedis";

export interface ConnectionOptions extends redis.ClientOpts {
    /**
     * Array of entity subscribers
     */
    entitySubscribers?: Array<{ new(...args: any[]): EntitySubscriberInterface<any> }>;
    /**
     * PubSub subscriber
     */
    pubSubSubscriber?: { new(...args: any[]): PubSubSubscriberInterface };
    
}

export class Connection {
    /**
     * Manager instance
     */
    public manager: RedisManager;
    
    /**
     * Connection status
     */
    public isConnected: boolean = false;
    
    /**
     * Redis client instance
     */
    public client: redis.RedisClient;
    
    /**
     * PubSub redis client
     * 
     * @private
     */
    private pubsub: redis.RedisClient;
    
    /**
     * Monitor redis client
     * 
     * @private
     */
    private monitor?: redis.RedisClient;
    
    /**
     * Options
     * 
     * @private
     */
    private options: ConnectionOptions;
    
    /**
     * Creates an instance of Connection.
     */
    public constructor() { }
    
    /**
     * Init connection & connect
     * 
     * @param options 
     */
    public async connect(options: ConnectionOptions): Promise<void> {
        if (this.isConnected) {
            throw new AlreadyConnectedError();
        }
        this.options = options;
        this.client = redis.createClient(options);
        this.pubsub = redis.createClient(options);
        try {
            this.initPubSubListener();
            const subscribers = this.loadEntitySubscribers(this.options.entitySubscribers);
            // tslint:disable-next-line:no-unused-expression
            subscribers;
            this.isConnected = true;
        } catch (e) {
            await this.disconnect();
            throw e;
        }
        // this.redisClient = redis.createClient();
        // await this.redisClient.connect();
        // try {
        //   const subscribers = this.loadSubscribers();
        //   this.manager = new RedisManager(this, subscribers)
        // } catch (e) {
        //
        // }
        // this.manager = new RedisManager();
        
    }

    /**
     * Disconnect all connections
     * 
     * @returns 
     */
    public async disconnect(): Promise<void> {
        // Stop monitor connection if exist
        if (this.monitor) {
            await this.monitor.quitAsync();
        }
        // Stop pubsub connection
        if (this.pubsub) {
            await this.pubsub.quitAsync();
        }
        // Stop client connection
        if (this.client) {
            await this.client.quitAsync();
        }
        this.isConnected = false;
    }
    
    /**
     * Creates monitor connection and executes MONITOR command
     */
    public async startMonitoring(logger: (time: number, args: any[], reply: string) => void): Promise<void> {
        if (this.monitor) {
            return;
        }
        this.monitor = redis.createClient(this.options);
        this.monitor.on("monitor", logger);
        await this.monitor.monitorAsync();
    }
    
    /**
     * Stops any monitoring
     * 
     * @returns 
     */
    public async stopMonitoring(): Promise<void> {
        if (this.monitor) {
            await this.monitor.quitAsync();
            this.monitor = undefined;
        }
    }
    
    /**
     * Subscribe for one or more chanels
     */
    public async subscribe(channel: string, ...channels: string[]): Promise<string> {
        return await this.pubsub.subscribeAsync(channel, ...channels);
    }
    
    /**
     * Stop listening for messages posted to the given channels.
     */
    public async unsubscribe(channel: string, ...channels: string[]): Promise<string> {
        return await this.pubsub.unsubscribeAsync(channel, ...channels);
    }

    /**
     * Stop listening for messages posted to channels matching the given patterns.
     */
    public async psubscribe(pattern: string, ...patterns: string[]): Promise<string> {
        return await this.pubsub.psubscribeAsync(pattern, ...patterns);
    }
    
    /**
     * Stop listening for messages posted to channels matching the given patterns.
     */
    public async punsubscribe(pattern: string, ...channels: string[]): Promise<string> {
        return await this.pubsub.punsubscribeAsync(pattern, ...channels);
    }
    
    /**
     * Remove all keys from the current database.
     */
    public async flushdb(): Promise<void> {
        await this.client.flushdbAsync();
    }
    
    /**
     * Execute multiple comamnds inside transaction
     * 
     * @param executor 
     * @param [atomic=false] 
     * @returns 
     */
    public async transaction(executor: (multi: redis.Commands<void>) => void, atomic: boolean = false): Promise<any[]> {
        const multi = this.client.multi();
        executor(multi);
        return new Promise<any[]>((resolve, reject) => {
            const func = atomic ? multi.exec_atomic : multi.exec;
            func.call(multi, (err: Error, result: any[]) => {
                err ? reject(err) : resolve(result);
            });
        });
    }
    
    /**
     * Execute multiple commands in batch without transaction
     * 
     * @param executor 
     * @returns 
     */
    public async batch(executor: (batch: redis.Commands<void>) => void): Promise<any[]> {
        const batch = this.client.batch();
        executor(batch);
        return new Promise<any[]>((resolve, reject) => {
            batch.exec((err, result) => {
                err ? reject(err) : resolve(result);
            });
        });
    }
    
    /**
     * Init pubsub connection and listeners
     * 
     * @private
     */
    private initPubSubListener(): void {
        if (typeof this.options.pubSubSubscriber === "function") {
            const listener = getFromContainer(this.options.pubSubSubscriber);
            if (listener.onMessage) {
                this.pubsub.on("message", listener.onMessage.bind(listener));
            }
            if (listener.onMessageBuffer) {
                this.pubsub.on("message_buffer", listener.onMessageBuffer.bind(listener));
            }
            if (listener.onPMessage) {
                this.pubsub.on("pmessage", listener.onPMessage.bind(listener));
            }
            if (listener.onPMessageBuffer) {
                this.pubsub.on("pmessage_buffer", listener.onPMessageBuffer.bind(listener));
            }
            if (listener.onSubscribe) {
                this.pubsub.on("subscribe", listener.onSubscribe.bind(listener));
            }
            if (listener.onPSubscribe) {
                this.pubsub.on("psubscribe", listener.onPSubscribe.bind(listener));
            }
            if (listener.onUnsubscribe) {
                this.pubsub.on("unsubscribe", listener.onUnsubscribe.bind(listener));
            }
            if (listener.onPUnsubscribe) {
                this.pubsub.on("punsubscribe", listener.onPUnsubscribe.bind(listener));
            }
        }
    }
    
    /**
     * Instantiate all entity subscribers
     * 
     * @private
     * @param subscriberConstructors 
     * @returns 
     */
    private loadEntitySubscribers(subscriberConstructors?: Array<{ new(...args: any[]): EntitySubscriberInterface<any> }>): Array<EntitySubscriberInterface<any>> {
        const subscribers: Array<EntitySubscriberInterface<any>> = [];
        if (subscriberConstructors) {
            for (const cl of subscriberConstructors) {
                subscribers.push(getFromContainer(cl));
            }
        }
        return subscribers;
    }
}