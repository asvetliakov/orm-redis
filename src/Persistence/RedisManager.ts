import { Connection } from "../Connection/Connection";
import { getEntityFullId, getEntityProperties, isRedisEntity } from "../Metadata/Metadata";
import { EntitySubscriberInterface } from "../Subscriber/EntitySubscriberInterface";
import { HydrationData, LoadOperation, Operator, PersistenceOperation } from "./Operator";

export type EntityType<T> = { new(): T } | Function;

/**
 * Main manager to get/save/remove entities
 * 
 * @export
 * @class RedisManager
 */
export class RedisManager {
    /**
     * Connection instance
     * 
     * @protected
     */
    public connection: Connection;

    /**
     * Array of entity subscribers
     * 
     * @protected
     */
    protected subscribers: Array<EntitySubscriberInterface<any>> = [];

    /**
     * Entity operator
     * 
     * @protected
     */
    protected operator: Operator;

    /**
     * Creates an instance of RedisManager.
     * @param connection 
     * @param subscribers 
     */
    public constructor(connection: Connection, subscribers: Array<EntitySubscriberInterface<any>>) {
        this.connection = connection;
        this.subscribers = subscribers;
        this.operator = new Operator();
    }
    
    /**
     * Save entity
     * 
     * @template T 
     * @param entity 
     * @returns 
     */
    public async save<T extends object>(entity: T): Promise<void> {
        // Run beforeSave subscribers for all entities regardless will they be finally saved or no
        // started from most deep relation entity to root entity
        const allEntities = this.getEntitiesForSubscribers(entity).reverse();
        for (const ent of allEntities) {
            const subscriber = this.subscribers.find(subscriber => subscriber.listenTo() === ent.constructor);
            if (subscriber && subscriber.beforeSave) {
                subscriber.beforeSave(ent);
            }
        }

        const operation = this.operator.getSaveOperation(entity);
        if (this.isEmptyPersistenceOperation(operation)) {
            return;
        }

        // Call entities subscribers
        // Do operation
        await this.connection.transaction(executor => {
            for (const deleteSet of operation.deletesSets) {
                executor.del(deleteSet);
            }
            for (const deleteHash of operation.deleteHashes) {
                executor.del(deleteHash);
            }
            for (const modifySet of operation.modifySets) {
                if (modifySet.removeValues.length > 0) {
                    executor.srem(modifySet.setName, modifySet.removeValues);
                }
                if (modifySet.addValues.length > 0) {
                    executor.sadd(modifySet.setName, modifySet.addValues);
                }
            }
            for (const modifyHash of operation.modifyHashes) {
                if (modifyHash.deleteKeys.length > 0) {
                    executor.hdel(modifyHash.hashId, modifyHash.deleteKeys);
                }
                if (Object.keys(modifyHash.changeKeys).length > 0) {
                    executor.hmset(modifyHash.hashId, modifyHash.changeKeys);
                }
            }
        });
        // update metadata
        this.operator.updateMetadataInHash(entity);
        // Call entities subscribers
        for (const ent of this.filterEntitiesForPersistenceOperation(allEntities, operation)) {
            const subscriber = this.subscribers.find(subscriber => subscriber.listenTo() === ent.constructor);
            if (subscriber && subscriber.afterSave) {
                subscriber.afterSave(ent);
            }
        }
    }

    
    /**
     * Get entity
     * 
     * @template T 
     * @param entityClass 
     * @param id 
     * @param [skipRelations] 
     * @returns 
     */
    public async load<T>(entityClass: EntityType<T>, id: string | number, skipRelations?: Array<keyof T>): Promise<T | undefined>;
    /**
     * Get entities
     * 
     * @template T 
     * @param entityClass 
     * @param id 
     * @param [skipRelations] 
     * @returns 
     */
    public async load<T>(entityClass: EntityType<T>, id: string[] | number[], skipRelations?: Array<keyof T>): Promise<T[] | undefined>;
    public async load<T>(entityClass: EntityType<T>, id: string | string[] | number | number[], skipRelations?: Array<keyof T>): Promise<T | undefined | T[]> {
        const idsToLoad = Array.isArray(id) ? id : [id];

        const entityIdToClassMap: Map<string, EntityType<any>> = new Map();
        const loadedData: Map<string, HydrationData> = new Map();
        const rootLoadOperations = idsToLoad.map(id => this.operator.getLoadOperation(id, entityClass, skipRelations)).filter(op => !!op) as LoadOperation[];

        rootLoadOperations.forEach(op => entityIdToClassMap.set(op.entityId, entityClass));

        const recursiveLoadDataWithRelations = async (operations: LoadOperation[]) => {
            const loadedDataForCall: HydrationData[] = [];

            // easier to use callbacks instead of processing result
            await this.connection.batch(executor => {
                for (const op of operations) {
                    executor.hgetall(op.entityId, (err, result) =>
                        loadedDataForCall.push({ id: op.entityId, redisData: result, entityClass: entityIdToClassMap.get(op.entityId) }));
                    for (const hash of op.hashes) {
                        executor.hgetall(hash, (err, result) => loadedDataForCall.push({ id: hash, redisData: result }));
                    }
                    for (const set of op.sets) {
                        executor.smembers(set, (err, result) => loadedDataForCall.push({ id: set, redisData: result }));
                    }
                }
            });
            loadedDataForCall.forEach(data => loadedData.set(data.id, data));
            const allMappings: LoadOperation["relationMappings"] = [];
            for (const op of operations) {
                allMappings.push(...op.relationMappings);
            }

            const relationOperations: Array<LoadOperation | undefined> = [];
            for (const mapping of allMappings) {
                const relationClass = mapping.relationClass;
                switch (mapping.type) {
                    // single relation in key
                    case "key": {
                        const hashVal = loadedDataForCall.find(data => data.id === mapping.ownerId);
                        if (hashVal && hashVal.redisData && !Array.isArray(hashVal.redisData) && hashVal.redisData[mapping.id]) {
                            const relationVal = hashVal.redisData[mapping.id];
                            if (relationVal && !loadedData.has(relationVal)) {
                                entityIdToClassMap.set(relationVal, relationClass);
                                relationOperations.push(this.operator.getLoadOperation(relationVal, relationClass));
                            }
                        }
                        break;
                    }
                    // set of relations
                    case "set": {
                        const set = loadedDataForCall.find(data => data.id === mapping.id);
                        if (set && set.redisData && Array.isArray(set.redisData)) {
                            for (const setVal of set.redisData) {
                                if (!loadedData.has(setVal)) {
                                    entityIdToClassMap.set(setVal, relationClass);
                                    relationOperations.push(this.operator.getLoadOperation(setVal, relationClass));
                                }
                            }
                        }
                        break;
                    }
                    // map of relations
                    case "map": {
                        const map = loadedDataForCall.find(data => data.id === mapping.id);
                        if (map && map.redisData && !Array.isArray(map.redisData) && Object.keys(map.redisData).length > 0) {
                            for (const key of Object.keys(map.redisData)) {
                                const relVal = map.redisData[key];
                                if (!loadedData.has(relVal)) {
                                    entityIdToClassMap.set(relVal, relationClass);
                                    relationOperations.push(this.operator.getLoadOperation(relVal, relationClass));
                                }
                            }
                        }
                        break;
                    }    
                }
            }
            if (relationOperations.length > 0) {
                await recursiveLoadDataWithRelations(relationOperations.filter(op => !!op) as LoadOperation[]);
            }
        };
        await recursiveLoadDataWithRelations(rootLoadOperations);
        const hydratedData = this.operator.hydrateData(loadedData);
        // run subscribers in reverse order
        hydratedData.reduceRight((unusued, data) => {
            if (data && data.constructor) {
                const subscriber = this.subscribers.find(sub => sub.listenTo() === data.constructor);
                if (subscriber && subscriber.afterLoad) {
                    subscriber.afterLoad(data);
                }
            }
        }, {});

        return Array.isArray(id)
            ? hydratedData.filter(data => data && data.constructor === entityClass)
            : hydratedData.find(data => data && data.constructor === entityClass);
    }

    /**
     * Remove entity. Doesn't remove linked relations
     * 
     * @param entity 
     * @returns 
     */
    public async remove(entity: object): Promise<void> { 
        const operation = this.operator.getDeleteOperation(entity);
        if (this.isEmptyPersistenceOperation(operation)) {
            return;
        }
        // Since we don't support cascade delete no need to process relations for subscribers
        const subscriber = this.subscribers.find(sub => sub.listenTo() === entity.constructor);
        if (subscriber && subscriber.beforeRemove) {
            subscriber.beforeRemove(entity);
        }
        await this.connection.transaction(executor => {
            for (const deleteSet of operation.deletesSets) {
                executor.del(deleteSet);
            }
            for (const deleteHash of operation.deleteHashes) {
                executor.del(deleteHash);
            }
        });
        this.operator.resetMetadataInEntityObject(entity);
        if (subscriber && subscriber.afterRemove) {
            subscriber.afterRemove(entity);
        }
    }

    /**
     * Serialize simple value to store it in redis
     * 
     * @param value 
     * @returns 
     */
    public serializeSimpleValue(value: any): string | undefined {
        return this.operator.serializeValue(value);
    }

    /**
     * Unserialize value
     * 
     * @param value 
     * @returns 
     */
    public unserializeSimpleValue(value: string): any {
        return this.operator.unserializeValue(value);
    }


    /**
     * Return all entities which should fire related subscribers for save/update/delete operation
     * 
     * @private
     * @param entity 
     * @param operation
     * @param [entities=[]] 
     * @returns 
     */
    private getEntitiesForSubscribers(entity: { [key: string]: any }, entities: object[] = []): object[] {
        if (!isRedisEntity(entity) || entities.includes(entity)) {
            return entities;
        }
        entities.push(entity);
        const metadata = getEntityProperties(entity); 
        if (!metadata) {
            return entities;
        }
        // const redisHash = getRedisHashId(entity);
        for (const propMetadata of metadata) {
            // no need to process non relations or without cascading options
            if (!propMetadata.isRelation || !(propMetadata.relationOptions.cascadeInsert || propMetadata.relationOptions.cascadeUpdate)) {
                continue;
            }
            const propValue = entity[propMetadata.propertyName];
            if (!propValue) {
                continue;
            }
            if (propValue instanceof Map || propValue instanceof Set) {
                for (const collVal of propValue.values()) {
                    this.getEntitiesForSubscribers(collVal, entities);
                }
            } else {
                this.getEntitiesForSubscribers(propValue, entities);
            }
        }
        return entities;
    }

    /**
     * Return entities which were somehow changed with given operation
     * 
     * @private
     * @param entities 
     * @param operation 
     * @returns 
     */
    private filterEntitiesForPersistenceOperation(entities: object[], operation: PersistenceOperation): object[] {
        return entities.filter(entity => {
            const hashId = getEntityFullId(entity);
            if (!hashId) {
                return false;
            }
            if (operation.deleteHashes.find(name => name.includes(hashId))) {
                return true;
            }
            if (operation.deletesSets.find(name => name.includes(hashId))) {
                return true;
            }
            if (operation.modifyHashes.find(v => ((v.deleteKeys.length > 0 || Object.keys(v.changeKeys).length > 0) && v.hashId.includes(hashId)))) {
                return true;
            }
            if (operation.modifySets.find(v => ((v.addValues.length > 0 || v.removeValues.length > 0) && v.setName.includes(hashId)))) {
                return true;
            }
            return false;
        });
    }

    /**
     * True true if persistence operation is empty (i.e. doesn't have any changes)
     * 
     * @private
     * @param operation 
     * @returns 
     */
    private isEmptyPersistenceOperation(operation: PersistenceOperation): boolean {
        return operation.deleteHashes.length === 0 && operation.deletesSets.length === 0 &&
            (operation.modifyHashes.length === 0 || operation.modifyHashes.every(val => val.deleteKeys.length === 0 && Object.keys(val.changeKeys).length === 0)) &&
            (operation.modifySets.length === 0 || operation.modifySets.every(val => val.addValues.length === 0 && val.removeValues.length === 0));
    }
}