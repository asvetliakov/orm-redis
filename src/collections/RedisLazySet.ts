import { getEntityFullId } from "../Metadata/Metadata";
import { EntityType, RedisManager } from "../Persistence/RedisManager";
import { LazySet } from "./LazySet";

/**
 * Redis-backed lazy set
 * 
 * @export
 * @class RedisLazySet
 * @template T 
 */
export class RedisLazySet<T> extends LazySet<T> {
    /**
     * Manager instance
     * 
     * @private
     */
    private manager: RedisManager;
    /**
     * Full set id
     * 
     * @private
     */
    private setId: string;

    /**
     * Relation class
     * 
     * @private
     */
    private entityClass?: Function;

    /**
     * Creates an instance of RedisLazySet.
     * @param setId Full set id in redis
     * @param manager Manager instance
     * @param [entityClass] If passed then set will be treated as entity set and will return entities
     */
    public constructor(setId: string, manager: RedisManager, entityClass?: EntityType<T>) {
        super();
        this.manager = manager;
        this.setId = setId;
        this.entityClass = entityClass;
    }

    /**
     * Add value or entity to the set
     * 
     * @param value 
     * @returns 
     */
    public async add(value: T): Promise<void> {
        if (this.entityClass) {
            // Entity
            await this.manager.save(value as any);
            await this.manager.connection.client.saddAsync(this.setId, getEntityFullId(value)!);
        } else {
            const serializedVal = this.manager.serializeSimpleValue(value);
            if (serializedVal) {
                await this.manager.connection.client.saddAsync(this.setId, serializedVal);
            }
        }
    }

    /**
     * Remove value or entity from the set. If set is the relation set then it's automatically remove entity
     * 
     * @param value 
     * @returns 
     */
    public async delete(value: T): Promise<boolean> {
        let res: number;
        if (this.entityClass) {
            // Entity
            await this.manager.remove(value as any);
            res = await this.manager.connection.client.sremAsync(this.setId, getEntityFullId(value)!);
        } else {
            const serializedVal = this.manager.serializeSimpleValue(value);
            if (serializedVal) {
                res = await this.manager.connection.client.sremAsync(this.setId, serializedVal);
            } else {
                res = 0;
            }
        }
        return res > 0 ? true : false;
    }

    /**
     * Determine if value or entity exists in the set
     * 
     * @param value 
     * @returns 
     */
    public async has(value: T): Promise<boolean> {
        const id = this.entityClass ? getEntityFullId(value)! : this.manager.serializeSimpleValue(value);
        if (typeof id === "undefined") {
            return false;
        }
        const res =  await this.manager.connection.client.sismemberAsync(this.setId, id);
        return !!res;
    }

    /**
     * Iterate over values
     */
    public async * values(): AsyncIterableIterator<T> {
        let [cursor, results]: [string, any[]] = await this.manager.connection.client.sscanAsync(this.setId, "0");
        // load entities
        if (this.entityClass && results.length > 0) {
            results = await this.manager.load(this.entityClass, results) as any;
        } else {
            results = results.map(res => this.manager.unserializeSimpleValue(res));
        }
        for (const res of results) {
            yield res;
        }
        while (cursor !== "0") {
            [cursor, results] = await this.manager.connection.client.sscanAsync(this.setId, cursor);
            // load entities
            if (this.entityClass && results.length > 0) {
                results = await this.manager.load(this.entityClass, results) as any;
            } else {
                results = results.map(res => this.manager.unserializeSimpleValue(res));
            }
            for (const res of results) {
                yield res;
            }
        }
    }

    /**
     * Iterator over values
     * 
     * @returns 
     */
    public async * [Symbol.asyncIterator](): AsyncIterableIterator<T> {
        for await (const val of this.values()) {
            yield val;
        }
    }
}