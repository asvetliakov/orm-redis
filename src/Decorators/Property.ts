import { LazyMap } from "../collections/LazyMap";
import { LazySet } from "../collections/LazySet";
import { PropertyMetadata, REDIS_PROPERTIES } from "../Metadata/Metadata";

export type ValidType =
    typeof Number |
    typeof String |
    typeof Boolean |
    typeof Object |
    typeof Array |
    typeof Date |
    typeof Map |
    typeof LazyMap |
    typeof LazySet |
    typeof Set;


/**
 * Defines redis hash property
 * 
 * @export
 * @param [type] 
 * @returns 
 */
export function Property(type?: ValidType): PropertyDecorator;
/**
 * Defines redis hash property
 * 
 * @export
 * @param name Redis name for property
 * @param [type] 
 * @returns 
 */
export function Property(name: string, type?: ValidType): PropertyDecorator; 

/**
 * Defines redis hash property
 * 
 * @export
 * @param [name] Property name to store in redis
 * @returns 
 */
export function Property(nameOrType?: string | ValidType, type?: ValidType): PropertyDecorator {
    return function (target: Object, propertyKey: string): void {
        const designType: any = Reflect.getMetadata("design:type", target, propertyKey);
        const name = typeof nameOrType === "string" ? nameOrType : propertyKey;
        const propertyType = typeof nameOrType === "function"
            ? nameOrType
            : type
                ? type
                : designType;        

        const properties: PropertyMetadata[] = Reflect.getMetadata(REDIS_PROPERTIES, target.constructor) || [];
        properties.push({
            propertyName: propertyKey,
            propertyRedisName: name,
            isIdentifyColumn: false,
            propertyType: propertyType,
            isRelation: false,
        });
        Reflect.defineMetadata(REDIS_PROPERTIES, properties, target.constructor);
    };
}