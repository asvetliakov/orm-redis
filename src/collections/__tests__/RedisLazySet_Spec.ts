import { Connection } from "../../Connection/Connection";
import { Entity } from "../../Decorators/Entity";
import { IdentifyProperty } from "../../Decorators/IdentifyProperty";
import { Property } from "../../Decorators/Property";
import { RedisManager } from "../../Persistence/RedisManager";
import { cleanRedisConnection, createRedisConnection } from "../../testutils/redis";
import { RedisTestMonitor } from "../../testutils/RedisTestMonitor";
import { RedisLazySet } from "../RedisLazySet";

let conn: Connection;
let manager: RedisManager;
let monitor: RedisTestMonitor;
beforeAll(async () => {
    conn = await createRedisConnection();
    manager = conn.manager;
    monitor = await RedisTestMonitor.create(conn);

});

afterEach(async () => {
    await monitor.clearMonitorCalls(50);
    await conn.flushdb();
});

afterAll(async () => {
    await monitor.release();
    await cleanRedisConnection(conn);
});

describe("Add", () => {
    it("Add simple value", async () => {
        const set = new RedisLazySet("a:mySet", manager);
        let res = await conn.client.smembersAsync("a:mySet");
        expect(res).toEqual([]);

        await set.add(1);
        await set.add(true);
        await set.add("test");
        res = await conn.client.smembersAsync("a:mySet");
        expect(res).toEqual(expect.arrayContaining([
            "i:1",
            "b:1",
            "s:test"
        ]));
    });

    it("Add entity", async () => {
        @Entity()
        class Ent {
            @IdentifyProperty()
            public id: number;

            @Property()
            public prop: string = "test prop";
        }

        const set = new RedisLazySet<Ent>("a:mySet", manager, Ent);
        const ent1 = new Ent();
        ent1.id = 1;
        const ent2 = new Ent();
        ent2.id = 2;
        await set.add(ent1);
        await set.add(ent2);

        const setRes = await conn.client.smembersAsync("a:mySet");
        expect(setRes).toEqual(expect.arrayContaining([
            "e:Ent:1",
            "e:Ent:2"
        ]));
        const res = await Promise.all([
            conn.client.hgetallAsync("e:Ent:1"),
            conn.client.hgetallAsync("e:Ent:2"),
        ]);
        expect(res).toMatchSnapshot();
    });
});

describe("Delete", () => {
    it("Delete simple value from set", async () => {
        await conn.client.saddAsync("a:mySet", "i:1", "i:2", "i:3");
        const set = new RedisLazySet("a:mySet", manager);
        await set.delete(1);
        await set.delete(2);
        await set.delete(10);

        const res = await conn.client.smembersAsync("a:mySet");
        expect(res).toEqual(["i:3"]);
    });

    it("Deletes entity from set and entity itself", async () => {
        @Entity()
        class Ent {
            @IdentifyProperty()
            public id: number;

            @Property()
            public prop: string = "test prop";
        }
        const set = new RedisLazySet<Ent>("a:mySet", manager, Ent);
        const ent1 = new Ent();
        ent1.id = 1;
        const ent2 = new Ent();
        ent2.id = 2;
        await manager.save(ent1);
        await manager.save(ent2);

        await set.add(ent1);
        await set.add(ent2);

        let setRes = await conn.client.smembersAsync("a:mySet");
        expect(setRes).toEqual(expect.arrayContaining([
            "e:Ent:1",
            "e:Ent:2"
        ]));

        await set.delete(ent2);
        setRes = await conn.client.smembersAsync("a:mySet");
        expect(setRes).toEqual([
            "e:Ent:1",
        ]);
        const ent2Res = await conn.client.hgetallAsync("e:Ent:2");
        expect(ent2Res).toBeNull();
    });
});

describe("Has", () => {
    it("Checks for existence of simple value", async () => {
        await conn.client.saddAsync("a:mySet", "i:1", "i:2", "i:3");
        const set = new RedisLazySet("a:mySet", manager);
        expect(await set.has(1)).toBeTruthy();
        expect(await set.has(2)).toBeTruthy();
        expect(await set.has("2")).toBeFalsy();
        expect(await set.has(10)).toBeFalsy();
    });

    it("Checks for existence of entity", async () => {
        @Entity()
        class Ent {
            @IdentifyProperty()
            public id: number;

            @Property()
            public prop: string = "test prop";
        }
        const set = new RedisLazySet<Ent>("a:mySet", manager, Ent);
        const ent1 = new Ent();
        ent1.id = 1;
        const ent2 = new Ent();
        ent2.id = 2;
        const ent3 = new Ent();
        ent3.id = 3;
        await set.add(ent1);
        await set.add(ent2);
        await manager.save(ent3);

        expect(await set.has(ent1)).toBeTruthy();
        expect(await set.has(ent2)).toBeTruthy();
        expect(await set.has(ent3)).toBeFalsy();
    });
});

describe("Values", () => {
    it("Iterates over simple set values", async () => {
        const prefill: number[] = [];
        for (let i = 0; i < 1000; i++) {
            prefill.push(i);
        }
        await conn.client.saddAsync("a:mySet", prefill.map(val => `i:${val}`));
        const set = new RedisLazySet<number>("a:mySet", manager);

        await monitor.clearMonitorCalls(150);

        const gettedValues: number[] = [];

        for await (const val of set.values()) {
            expect(typeof val).toBe("number");
            gettedValues.push(val);
        }

        expect(gettedValues).toHaveLength(1000);
        expect(gettedValues).toEqual(expect.arrayContaining(prefill));
        await monitor.wait(100);
        // req[0] is sscan
        // req[1] is a:mySet
        expect(monitor.requests.map(req => [req[0], [req[1]]]).length).toBeGreaterThan(10);
    });

    it("Iterates over entities", async () => {
        @Entity()
        class Ent {
            @IdentifyProperty()
            public id: number;

            @Property()
            public prop: string = "test prop";
        }
        const idsInSet: string[] = [];
        for (let i = 0; i < 50; i++) {
            const ent = new Ent();
            ent.id = i;
            await manager.save(ent);
            idsInSet.push(`e:Ent:${i}`);
        }
        await conn.client.saddAsync("a:mySet", idsInSet);
        const set = new RedisLazySet<Ent>("a:mySet", manager, Ent);

        await monitor.clearMonitorCalls(150);
        const entites: Ent[] = [];
        for await (const val of set.values()) {
            entites.push(val);
        }
        expect(entites).toHaveLength(50);
        expect(entites[0]).toBeInstanceOf(Ent);
        expect(entites.map(ent => ent.id)).toEqual(expect.arrayContaining([
            0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
            30, 40, 41, 42, 43, 44, 45, 49 
        ]));
        await monitor.wait(100);
        // req[0] is sscan
        // req[1] is a:mySet
        expect(monitor.requests.map(req => [req[0], [req[1]]]).length).toBeGreaterThan(10);
    });
});