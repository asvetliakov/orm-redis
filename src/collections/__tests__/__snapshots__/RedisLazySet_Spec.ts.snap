// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Add entity 1`] = `
Array [
  Object {
    "id": "i:1",
    "prop": "s:test prop",
  },
  Object {
    "id": "i:2",
    "prop": "s:test prop",
  },
]
`;
